http-client/src/main/java/io/micronaut/http/client/DefaultHttpClient.java

```diff
@@ -1838,12 +1838,18 @@ protected void channelRead0(ChannelHandlerContext channelHandlerContext, FullHtt
                             // normalize the NO_CONTENT header, since http content aggregator adds it even if not present in the response
                             headers.remove(HttpHeaderNames.CONTENT_LENGTH);
                         }
-                        boolean errorStatus = statusCode >= 400;
+
+                        boolean convertBodyWithBodyType = statusCode < 400 ||
+                                (!DefaultHttpClient.this.configuration.isExceptionOnErrorStatus() && bodyType.equalsType(errorType));
```

Comment:
Can you explain the bodyType equals error type? Why is that part of the equation?

Reply:
Usage example:
```
HttpResponse<MyBody> res = httpClient.toBlocking().exchange(request, MyBody.class);
// res.body(); -- returns `MyBody` type
```
and errorType defaults to `JsonError`.
`exchange` can only _return_ the HttpResponse when both types match, i.e when the type on the left-hand side of the equal (i.e bodyType) matches with the type returned in case of error. 
Hence this check.

Reply:
@jameskleeh Anything else I should do to get this merged in?